### PR TITLE
sync: Add RwLock::{try_read, try_write}

### LINKED ISSUE #2284: RwLock miss try_read and try_write method
## Version 0.2.13

> ```
> └── tokio v0.2.13
>     └── tokio-macros v0.2.5
> ```

## Platform

archlinux x64

## Description

I try 
```rust
use tokio::sync::RwLock;

#[tokio::main]
async fn main() {
    let rw_lock = RwLock::new(1);
    
    rw_lock.try_read();
}
```

It should compile well. But cargo report

> method not found in `tokio::sync::rwlock::RwLock<{integer}>`

I read the doc, RwLock doesn't have `try_read` or `try_write`, if I want to try to acquire a read lock or write lock but don't block when acquire failed, what should I do

### PR DESCRIPTION
**Pull request by pingw33n**
Closes #2284

### CODE REVIEW

File: tokio/src/sync/tests/loom_rwlock.rs [hunk 1]
```diff
@@ -76,3 +119,39 @@ fn concurrent_read_write() {
...
+        let rwclone = rwlock.clone();
+        let t2 = thread::spawn(move || {
+            let mut guard = spin_write(&rwclone);
+            *guard += 5;
+        });
+
+        let rwclone = rwlock.clone();
+        let t3 = thread::spawn(move || {
+            let guard = spin_read(&rwclone);
+            //at this state the value on the lock may either be 0, 5, or 10
```
Comment (hawkw) [inline tokio/src/sync/tests/loom_rwlock.rs:145]: nit: prefer a space after the comment
```suggestion
            // at this state the value on the lock may either be 0, 5, or 10
```

File: tokio/src/sync/tests/loom_rwlock.rs [hunk 2]
```diff
@@ -76,3 +119,39 @@ fn concurrent_read_write() {
...
+            assert!(*guard == 0 || *guard == 5 || *guard == 10);
+        });
+
+        t1.join().expect("thread 1 write should not panic");
+        t2.join().expect("thread 2 write should not panic");
+        t3.join().expect("thread 3 read should not panic");
+
+        let guard = spin_read(&rwlock);
+        //when all threads have finished the value on the lock should be 10
```
Comment (hawkw) [inline tokio/src/sync/tests/loom_rwlock.rs:154]: nit:
```suggestion
        // when all threads have finished the value on the lock should be 10
```

File: tokio/src/sync/tests/loom_rwlock.rs [hunk 3]
```diff
@@ -35,6 +35,49 @@ fn concurrent_write() {
...
+
+        let rwclone = rwlock.clone();
+        let t2 = thread::spawn(move || {
+            let mut guard = spin_write(&rwclone);
+            *guard += 5;
+        });
+
+        t1.join().expect("thread 1 write should not panic");
+        t2.join().expect("thread 2 write should not panic");
+        //when all threads have finished the value on the lock should be 10
```
Comment (hawkw) [inline tokio/src/sync/tests/loom_rwlock.rs:75]: nit: 
```suggestion
        // when all threads have finished the value on the lock should be 10
```

**Comment by carllerche** [comment]
Looks like master needs to be merged.

**Comment by Darksonn** [comment]
I merged master, and it broke with the following message:
```
error[E0308]: mismatched types
   --> tokio/src/sync/rwlock.rs:127:28
    |
127 |         lock.s.try_acquire(num_permits)?;
    |                            ^^^^^^^^^^^
    |                            |
    |                            expected `u32`, found `u16`
    |                            help: you can convert an `u16` to `u32`: `num_permits.into()`
```

**Comment by carllerche** [comment]
Closing due to inactivity. A new PR can be opened tracking master.

**Pull request closed without merging**